From bf82c200bfb31353dcc3ae2786bc8cb664dc9286 Mon Sep 17 00:00:00 2001
From: Manikanta Pubbisetty <mpubbise@codeaurora.org>
Date: Fri, 2 Aug 2019 10:41:20 +0530
Subject: [PATCH 2/2] ath11k: enable pn offload

Enabling PN checking in the hardware; i.e, the hardware checks
the PN of the received packets and report the error to
the host for further handling. PN errors will be reported via
WBM rx release ring and the default action is to drop the
packet with PN error.

The following command can be used to check for PN errors;
"cat /sys/kernel/debug/ath11k/soc_rx_stats | grep -i pn"

Sample Output:
PN check fail: 210
PN err: 0

Signed-off-by: Manikanta Pubbisetty <mpubbise@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/core.h   |   1 +
 drivers/net/wireless/ath/ath11k/dp.c     |   4 +-
 drivers/net/wireless/ath/ath11k/dp_rx.c  | 143 ++++++++++++++++++++++++++++---
 drivers/net/wireless/ath/ath11k/dp_rx.h  |   7 +-
 drivers/net/wireless/ath/ath11k/hal.h    |   2 +-
 drivers/net/wireless/ath/ath11k/hal_rx.c |  14 ++-
 drivers/net/wireless/ath/ath11k/mac.c    |  31 +++++++
 7 files changed, 185 insertions(+), 17 deletions(-)

Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/core.h
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/core.h
@@ -363,6 +363,7 @@ struct ath11k_sta {
 	u32 bw;
 	u32 nss;
 	u32 smps;
+	enum hal_pn_type pn_type;
 
 	struct work_struct update_wk;
 	struct rate_info txrate;
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp.c
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.c
@@ -58,14 +58,14 @@ int ath11k_dp_peer_setup(struct ath11k *
 	}
 
 	ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id,
-				       HAL_DESC_REO_NON_QOS_TID, 1, 0);
+				       HAL_DESC_REO_NON_QOS_TID, 1, 0, HAL_PN_TYPE_NONE);
 	if (ret) {
 		ath11k_warn(ab, "failed to setup rxd tid queue for non-qos tid %d\n",
 			    ret);
 		return ret;
 	}
 
-	ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id, 0, 1, 0);
+	ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id, 0, 1, 0, HAL_PN_TYPE_NONE);
 	if (ret) {
 		ath11k_warn(ab, "failed to setup rxd tid queue for tid 0 %d\n",
 			    ret);
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -843,7 +843,8 @@ unlock_exit:
 }
 
 int ath11k_peer_rx_tid_setup(struct ath11k *ar, const u8 *peer_mac, int vdev_id,
-			     u8 tid, u32 ba_win_sz, u16 ssn)
+			     u8 tid, u32 ba_win_sz, u16 ssn,
+			     enum hal_pn_type pn_type)
 {
 	struct ath11k_base *ab = ar->ab;
 	struct ath11k_peer *peer;
@@ -904,7 +905,8 @@ int ath11k_peer_rx_tid_setup(struct ath1
 
 	addr_aligned = PTR_ALIGN(vaddr, HAL_LINK_DESC_ALIGN);
 
-	ath11k_hal_reo_qdesc_setup(addr_aligned, tid, ba_win_sz, ssn);
+	ath11k_hal_reo_qdesc_setup(addr_aligned, tid, ba_win_sz,
+				   ssn, pn_type);
 
 	paddr = dma_map_single(ab->dev, addr_aligned, hw_desc_sz,
 			       DMA_BIDIRECTIONAL);
@@ -948,7 +950,7 @@ int ath11k_dp_rx_ampdu_start(struct ath1
 
 	ret = ath11k_peer_rx_tid_setup(ar, params->sta->addr, vdev_id,
 				       params->tid, params->buf_size,
-				       params->ssn);
+				       params->ssn, arsta->pn_type);
 	if (ret)
 		ath11k_warn(ab, "failed to setup rx tid %d\n", ret);
 
@@ -1001,8 +1003,80 @@ int ath11k_dp_rx_ampdu_stop(struct ath11
 	return ret;
 }
 
-static int ath11k_get_ppdu_user_index(struct htt_ppdu_stats *ppdu_stats,
-				      u16 peer_id)
+int ath11k_dp_peer_rx_pn_replay_config(struct ath11k_vif *arvif,
+				       const u8 *peer_addr,
+				       enum set_key_cmd key_cmd,
+				       struct ieee80211_key_conf *key)
+{
+	struct ath11k *ar = arvif->ar;
+	struct ath11k_base *ab = ar->ab;
+	struct ath11k_hal_reo_cmd cmd = {0};
+	struct ath11k_peer *peer;
+	struct dp_rx_tid *rx_tid;
+	u8 tid;
+	int ret = 0;
+
+	/* NOTE: Enable PN/TSC replay check offload only for unicast frames.
+	 * We use mac80211 PN/TSC replay check functionality for bcast/mcast
+	 * for now.
+	 */
+	if (!(key->flags & IEEE80211_KEY_FLAG_PAIRWISE))
+		return 0;
+
+	cmd.flag |= HAL_REO_CMD_FLG_NEED_STATUS;
+	cmd.upd0 |= HAL_REO_CMD_UPD0_PN |
+		    HAL_REO_CMD_UPD0_PN_SIZE |
+		    HAL_REO_CMD_UPD0_PN_VALID |
+		    HAL_REO_CMD_UPD0_PN_CHECK |
+		    HAL_REO_CMD_UPD0_SVLD;
+
+	switch (key->cipher) {
+	case WLAN_CIPHER_SUITE_TKIP:
+	case WLAN_CIPHER_SUITE_CCMP:
+	case WLAN_CIPHER_SUITE_CCMP_256:
+	case WLAN_CIPHER_SUITE_GCMP:
+	case WLAN_CIPHER_SUITE_GCMP_256:
+		if (key_cmd == SET_KEY) {
+			cmd.upd1 |= HAL_REO_CMD_UPD1_PN_CHECK;
+			cmd.pn_size = 48;
+		}
+		break;
+	default:
+		break;
+	}
+
+	spin_lock_bh(&ab->base_lock);
+
+	peer = ath11k_peer_find(ab, arvif->vdev_id, peer_addr);
+	if (!peer) {
+		ath11k_warn(ab, "failed to find the peer to configure pn replay detection\n");
+		spin_unlock_bh(&ab->base_lock);
+		return -ENOENT;
+	}
+
+	for (tid = 0; tid <= IEEE80211_NUM_TIDS; tid++) {
+		rx_tid = &peer->rx_tid[tid];
+		if (!rx_tid->active)
+			continue;
+		cmd.addr_lo = lower_32_bits(rx_tid->paddr);
+		cmd.addr_hi = upper_32_bits(rx_tid->paddr);
+		ret = ath11k_dp_tx_send_reo_cmd(ab, rx_tid,
+						HAL_REO_CMD_UPDATE_RX_QUEUE,
+						&cmd, NULL);
+		if (ret) {
+			ath11k_warn(ab, "failed to configure rx tid %d queue for pn replay detection %d\n",
+				    tid, ret);
+			break;
+		}
+	}
+
+	spin_unlock_bh(&ar->ab->base_lock);
+
+	return ret;
+}
+
+static inline int ath11k_get_ppdu_user_index(struct htt_ppdu_stats *ppdu_stats,
+					     u16 peer_id)
 {
 	int i;
 
@@ -2092,7 +2166,7 @@ static void ath11k_dp_rx_h_mpdu(struct a
 	struct sk_buff *last_msdu;
 	struct sk_buff *msdu;
 	struct ath11k_skb_rxcb *last_rxcb;
-	bool is_decrypted = false;
+	bool is_decrypted = false, fill_crypto_hdr;
 	u32 err_bitmap;
 	u8 *qos;
 
@@ -2115,6 +2189,9 @@ static void ath11k_dp_rx_h_mpdu(struct a
 
 	err_bitmap = ath11k_dp_rx_h_attn_mpdu_err(last_rxcb->rx_desc);
 
+	/* PN for multicast packets will be checked in mac80211 */
+	fill_crypto_hdr = is_multicast_ether_addr(hdr->addr1);
+
 	enctype = ath11k_dp_rx_h_mpdu_start_enctype(rx_desc);
 	if (enctype != HAL_ENCRYPT_TYPE_OPEN && !err_bitmap)
 		is_decrypted = ath11k_dp_rx_h_attn_is_decrypted(rx_desc);
@@ -2132,14 +2209,27 @@ static void ath11k_dp_rx_h_mpdu(struct a
 	if (err_bitmap & DP_RX_MPDU_ERR_TKIP_MIC)
 		rx_status->flag |= RX_FLAG_MMIC_ERROR;
 
-	if (is_decrypted)
-		rx_status->flag |= RX_FLAG_DECRYPTED | RX_FLAG_MMIC_STRIPPED |
-				   RX_FLAG_MIC_STRIPPED | RX_FLAG_ICV_STRIPPED;
+	if (is_decrypted) {
+		rx_status->flag |= RX_FLAG_DECRYPTED | RX_FLAG_MMIC_STRIPPED;
+
+		if (fill_crypto_hdr)
+			rx_status->flag |= RX_FLAG_MIC_STRIPPED |
+					RX_FLAG_ICV_STRIPPED;
+		else
+			rx_status->flag |= RX_FLAG_IV_STRIPPED |
+					   RX_FLAG_PN_VALIDATED;
+	}
 
 	skb_queue_walk(amsdu_list, msdu) {
 		ath11k_dp_rx_h_csum_offload(msdu);
 		ath11k_dp_rx_h_undecap(ar, msdu, rx_desc,
 				       enctype, rx_status, is_decrypted);
+
+		if (!is_decrypted || fill_crypto_hdr)
+			continue;
+
+		hdr = (void *)msdu->data;
+		hdr->frame_control &= ~__cpu_to_le16(IEEE80211_FCTL_PROTECTED);
 	}
 }
 
@@ -3735,6 +3825,7 @@ static int ath11k_dp_rx_h_null_q_desc(st
 	struct dp_rx_tid *rx_tid;
 	u16 msdu_len, peer_id;
 	u8 l3pad_bytes, tid;
+	enum hal_pn_type pn_type;
 	u8 peer_mac[ETH_ALEN];
 	int vdev_id;
 
@@ -3800,10 +3891,11 @@ static int ath11k_dp_rx_h_null_q_desc(st
 
 		if (!rx_tid->active) {
 			vdev_id = peer->vdev_id;
+			pn_type = arsta->pn_type;
 			ether_addr_copy(peer_mac, peer->addr);
 			spin_unlock_bh(&ar->ab->base_lock);
 			ath11k_peer_rx_tid_setup(ar, peer_mac, vdev_id,
-						 tid, 1, 0);
+						 tid, 1, 0, pn_type);
 			spin_lock_bh(&ar->ab->base_lock);
 		}
 	}
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.h
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.h
@@ -43,9 +43,14 @@ int ath11k_dp_rx_ampdu_start(struct ath1
 			     struct ieee80211_ampdu_params *params);
 int ath11k_dp_rx_ampdu_stop(struct ath11k *ar,
 			    struct ieee80211_ampdu_params *params);
+int ath11k_dp_peer_rx_pn_replay_config(struct ath11k_vif *arvif,
+				       const u8 *peer_addr,
+				       enum set_key_cmd key_cmd,
+				       struct ieee80211_key_conf *key);
 void ath11k_peer_rx_tid_cleanup(struct ath11k *ar, struct ath11k_peer *peer);
 int ath11k_peer_rx_tid_setup(struct ath11k *ar, const u8 *peer_mac, int vdev_id,
-			     u8 tid, u32 ba_win_sz, u16 ssn);
+			     u8 tid, u32 ba_win_sz, u16 ssn,
+			     enum hal_pn_type pn_type);
 void ath11k_dp_htt_htc_t2h_msg_handler(struct ath11k_base *ab,
 				       struct sk_buff *skb);
 int ath11k_dp_pdev_reo_setup(struct ath11k_base *ab);
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/hal.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/hal.h
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/hal.h
@@ -847,7 +847,7 @@ struct ath11k_hal {
 
 u32 ath11k_hal_reo_qdesc_size(u32 ba_window_size, u8 tid);
 void ath11k_hal_reo_qdesc_setup(void *vaddr, int tid, u32 ba_window_size,
-				u32 start_seqtype);
+				u32 start_seq, enum hal_pn_type type);
 void ath11k_hal_reo_init_cmd_ring(struct ath11k_base *ab,
 				  struct hal_srng *srng);
 void ath11k_hal_reo_hw_setup(struct ath11k_base *ab);
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/hal_rx.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/hal_rx.c
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/hal_rx.c
@@ -735,7 +735,7 @@ u32 ath11k_hal_reo_qdesc_size(u32 ba_win
 }
 
 void ath11k_hal_reo_qdesc_setup(void *vaddr, int tid, u32 ba_window_size,
-				u32 start_seq)
+				u32 start_seq, enum hal_pn_type type)
 {
 	struct hal_rx_reo_queue *qdesc = (struct hal_rx_reo_queue *)vaddr;
 	struct hal_rx_reo_queue_ext *ext_desc;
@@ -764,6 +764,18 @@ void ath11k_hal_reo_qdesc_setup(void *va
 
 	qdesc->info0 |= FIELD_PREP(HAL_RX_REO_QUEUE_INFO0_BA_WINDOW_SIZE,
 				   ba_window_size - 1);
+	switch (type) {
+	case HAL_PN_TYPE_NONE:
+	case HAL_PN_TYPE_WAPI_EVEN:
+	case HAL_PN_TYPE_WAPI_UNEVEN:
+		break;
+	case HAL_PN_TYPE_WPA:
+		qdesc->info0 |=
+			FIELD_PREP(HAL_RX_REO_QUEUE_INFO0_PN_CHECK, 1) |
+			FIELD_PREP(HAL_RX_REO_QUEUE_INFO0_PN_SIZE,
+				   HAL_RX_REO_QUEUE_PN_SIZE_48);
+		break;
+	}
 
 	/* TODO: Set Ignore ampdu flags based on BA window size and/or
 	 * AMPDU capabilities
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/mac.c
@@ -2310,6 +2310,7 @@ static int ath11k_mac_op_set_key(struct
 	struct ath11k_base *ab = ar->ab;
 	struct ath11k_vif *arvif = ath11k_vif_to_arvif(vif);
 	struct ath11k_peer *peer;
+	struct ath11k_sta *arsta;
 	const u8 *peer_addr;
 	int ret = 0;
 	u32 flags = 0;
@@ -2370,6 +2371,13 @@ static int ath11k_mac_op_set_key(struct
 		goto exit;
 	}
 
+	ret = ath11k_dp_peer_rx_pn_replay_config(arvif, peer_addr, cmd, key);
+	if (ret) {
+		ath11k_warn(ab, "failed to enable PN replay detection offload support %d\n",
+			   ret);
+		goto exit;
+	}
+
 	spin_lock_bh(&ab->base_lock);
 	peer = ath11k_peer_find(ab, arvif->vdev_id, peer_addr);
 	if (peer && cmd == SET_KEY) {
@@ -2387,6 +2395,26 @@ static int ath11k_mac_op_set_key(struct
 	} else if (!peer)
 		/* impossible unless FW goes crazy */
 		ath11k_warn(ab, "peer %pM disappeared!\n", peer_addr);
+
+	if (sta) {
+		arsta = (struct ath11k_sta *)sta->drv_priv;
+
+		switch (key->cipher) {
+		case WLAN_CIPHER_SUITE_TKIP:
+		case WLAN_CIPHER_SUITE_CCMP:
+		case WLAN_CIPHER_SUITE_CCMP_256:
+		case WLAN_CIPHER_SUITE_GCMP:
+		case WLAN_CIPHER_SUITE_GCMP_256:
+			if (cmd == SET_KEY)
+				arsta->pn_type = HAL_PN_TYPE_WPA;
+			else
+				arsta->pn_type = HAL_PN_TYPE_NONE;
+			break;
+		default:
+			arsta->pn_type = HAL_PN_TYPE_NONE;
+			break;
+		}
+	}
 	spin_unlock_bh(&ab->base_lock);
 
 exit:
